### PR TITLE
Update to Mol* v3, support for pLDDT coloring

### PR DESCRIPTION
**Pull request by sbittrich**
This provides some housekeeping and makes the project work with the most recent Mol* version. Additionally, there's now support to color chains by pLDDT values.

Still wip. The one thing I couldn't get to work is lighting. I didn't spot a difference config- or property-wise, yet everything renders without light. Do you have any pointers to what I'm doing wrong?

Also, the expression for `ligandPlusConnected` changed. See e.g. [this](https://www.rcsb.org/3d-view/1G2F?preset=ligandInteraction&label_asym_id=G), select the ion, then manipulate the selection to include residues with metallic bonds. Previously, the coordinating residues were selected (and showed up in images), but I think this change is intentional.

Current "lighting":
ma-bak-cepc-0001_model-1

With `ignoreLight`:
ma-bak-cepc-0001_chain-B

Thanks for having a look!

### CODE REVIEW

File: src/render.ts [hunk 1]
```diff
@@ -164,408 +185,288 @@ export class ImageRenderer {
                 },
                 outline: {
                     name: 'off', params: {}
+                },
+                antialiasing: {
+                    name: 'off', params: {}
```
Comment (arose) [inline src/render.ts]: I would add SMAA

File: src/render.ts [hunk 2]
```diff
@@ -164,408 +185,288 @@ export class ImageRenderer {
...
-            drawPass: {
-                cameraHelper: {
-                    axes: { name: 'off', params: {} }
-                }
+            cameraHelper: {
+                axes: { name: 'off', params: {} }
             },
             multiSample: {
                 mode: 'on',
                 sampleLevel: 3
```
Comment (arose) [inline src/render.ts]: I would use 4

File: src/render.ts [hunk 3]
```diff
@@ -164,408 +185,288 @@ export class ImageRenderer {
...
      * Creates PNG with the current 3dcanvas data
      */
     async createImage(outPath: string, size: StructureSize) {
         const occlusion = size === StructureSize.Big ? { name: 'on' as const, params: {
-            kernelSize: 4,
-            bias: 0.5,
+            samples: 32,
             radius: 64,
-        } } : { name: 'off' as const, params: {} }
+            bias: 0.5,
```
Comment (arose) [inline src/render.ts]: The meaning of those values has changed, the current defaults are

```
{
  samples: 32,
  radius: 5,
  bias: 0.8,
  blurKernelSize: 15,
  resolutionScale: 1
}
```

File: src/render.ts [hunk 4]
```diff
@@ -164,408 +185,288 @@ export class ImageRenderer {
...
+            blurKernelSize: 4,
+            resolutionScale: 1,
+        } } : { name: 'off' as const, params: {} };
         const outline = size === StructureSize.Big ? { name: 'on' as const, params: {
             scale: 1,
```
Comment (arose) [inline src/render.ts:300]: threshold should be 0.33, again the meaning has chnaged

File: src/render.ts [hunk 5]
```diff
@@ -130,29 +145,35 @@ interface ReprParams {
...
-        this.gl = getGLContext(this.width, this.height, {
+    webgl: WebGLContext;
+    reprCtx: RepresentationContext;
+    canvas3d: Canvas3D;
+    imagePass: ImagePass;
+    assetManager = new AssetManager();
+
+    constructor(private width: number, private height: number, private format: 'png' | 'jpeg', private plddt: 'on' | 'single-chain' | 'off') {
+        this.webgl = createContext(getGLContext(this.width, this.height, {
             alpha: false,
```
Comment (arose) [inline src/render.ts]: use
```
{
  antialias: true,
  preserveDrawingBuffer: true,
  alpha: true,  // the renderer requires an alpha channel
  depth: true,  // the renderer requires a depth buffer
  premultipliedAlpha: true // the renderer outputs PMA
}
```

**Review by arose — COMMENTED** [review]
Not sure what is going on with the lights. Maybe it gets fixed by some of the suggested param changes.

Generally style/material params have moved from the renderer params to each representation params.

File: src/render.ts (hunk 1 above)
Comment (sbittrich) [reply src/render.ts]: This results in `Error: SMAA not supported, missing "HTMLImageElement"` due to running in headless mode. I assumed that also the reason why AA was disabled in the original code.

File: src/render.ts (hunk 2 above)
Comment (sbittrich) [reply src/render.ts]: I've updated all params to your recommendations.

File: src/render.ts (hunk 3 above)
Comment (sbittrich) [reply src/render.ts]: Updated.

File: src/render.ts (hunk 4 above)
Comment (sbittrich) [reply src/render.ts:300]: Updated.

File: src/render.ts (hunk 5 above)
Comment (sbittrich) [reply src/render.ts]: Updated.

**Comment by sbittrich** [comment]
> Not sure what is going on with the lights. Maybe it gets fixed by some of the suggested param changes.
> 
> Generally style/material params have moved from the renderer params to each representation params.

Thanks for the param recommendations. Unfortunately, the lighting issues persist.

Is this maybe a problem in headless-gl? I'm seeing that a number of WebGL extensions aren't supported.

```
Could not find support for "texture_half_float"
Could not find support for "texture_half_float_linear"
Could not find support for "depth_texture"
Could not find support for "frag_depth"
error creating framebuffer for '36193'
Could not find support for "color_buffer_half_float"
Could not find support for "shader_texture_lod"
Could not find support for "sRGB"
Missing "depthTexture" extensions required for "wboit"
Missing "HTMLImageElement" required for "SMAA"
Draw error rendering item id 0: 'Error: WebGL error: 'invalid enum''
```

I'll try to investigate some more. Just wanted to make sure that I'm not missing something obvious with the v3 lighting changes.

**Comment by sbittrich** [comment]
Nevermind, it's none of the debug statements. Lighting broke with the introduction of multiple lights.

The issue is in headless-gl, which requires a `[0]` suffix when getting the location of array uniforms (https://github.com/stackgl/headless-gl/issues/170). Without neither `uLightDirection` nor `uLightColor` get set.

Conditionally appending the name by `[0]` in https://github.com/molstar/molstar/blob/4d97ccdfb336011838dcc9ba90da318180112c35/src/mol-gl/webgl/program.ts#L45 works.

Kinda ugly, but I don't think this will get fixed in headless-gl directly.

File: src/render.ts (hunk 1 above)
Comment (arose) [reply src/render.ts]: right, in that case you can try fxaa and see if it help a bit

**Comment by sbittrich** [comment]
Lightning is now fixed, thanks to the new Mol* release.
ma_mabakcepc0001_model-1
ma_mabakcepc0001_chain-B

I've also tweaked the occlusion params to resemble more closely what people are used to.

0.33 threshold:
3j3q_assembly-1-0 33

0.95 threshold:
3j3q_assembly-1-0 95

Current image:
3j3q_assembly-1

This should address all issues. We can think about adding other things later on.

File: src/render.ts (hunk 1 above)
Comment (sbittrich) [reply src/render.ts]: Switched to fxaa with [acf9b1c](https://github.com/molstar/molrender/pull/3/commits/acf9b1cf289c4df785f8ec57151a40e571cb6649).

**Comment by sbittrich** [comment]
@arose Can this be merged & released? I only saw after opening the PR that I have write access to this repo, but either way I probably can't release to npm.

**Review by arose — APPROVED** [review]

**Comment by arose** [comment]
@JonStargaryen fine by me. I also invited you on npm as a maintainer for this.

**Comment by sbittrich** [comment]
Thanks!

**Pull request merged**